### PR TITLE
Add gpt baselines: mini, small, medium

### PR DESCRIPTION
**Pull request by TomaszOdrzygozdz**
# Description
todo

# Checklist
Answer each question with "yes", and provide an explanation if the answer isn't what is expected.

1. Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - Answer: yes
2. Are all variables set with logical defaults that are backward compatible?
   - Answer: yes
3. Can you provide a link to a Neptune experiment verifying the functionality of this PR?
   - Answer: N/A
4. Have you successfully executed the linter and tests?
   - Answer: yes
5. Are all functions concise and don't require splitting?
   - Answer: yes
6. Is there documentation for arguments and complex logic?
   - Answer: N/A
7. Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - Answer: yes
8. Are all functions placed in appropriate files?
   - Answer: N/A
9. Is the PR name meaningful and descriptive?
   - Answer: yes
10. Is PR description provided, or unnecessary?
    - Answer: unnecessary
  
Other questions:

11. Have you not made changes to the requirements or anything related to the Singularity image? If you have, did you generate and upload a new image to the clusters?
    - Answer: yes
12. Is the change significant enough to notify all individuals working with this repository?
    - Answer: yes
13. Is there a need for a second reviewer?
    - Answer: yes

### CODE REVIEW

File: research/conditional/train/configs/baselines/bert_medium.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# grid args
+runner: "research.conditional.train.cc_train"
+time: "0-02:00:00"
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/bert_medium.yaml:3]: czy 2h serio wystarczają?

File: research/conditional/train/configs/baselines/gpt_small.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  save_weights_interval: 1000
+
+  use_neptune: true
+  project_name: "pmtest/llm-random"
+  name: "gpt_medium_baseline"
+  mixed_precision: true
+  tags:
+    - "gpt"
+    - "baseline"
+    - "medium"
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/gpt_small.yaml:38]: tag: small

File: research/conditional/train/configs/baselines/bert_mini.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  save_weights_path: "weights"
+  save_weights_interval: 1000
+
+  use_neptune: true
+  project_name: "pmtest/llm-random"
+  name: "gpt_mini_baseline"
+  mixed_precision: true
+  tags:
+    - "bert"
+    - "baseline"
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/bert_mini.yaml:38]: wgl może "densebaseline" lub "densebs" for short?

File: research/conditional/train/configs/baselines/bert_medium.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  logging_interval_loss: 1000
+  save_weights_path: "weights"
+  save_weights_interval: 1000
+
+  use_neptune: true
+  project_name: "pmtest/llm-random"
+  name: "gpt_medium_baseline"
+  mixed_precision: true
+  tags:
+    - "bert"
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/bert_medium.yaml:37]: gpt, nie bert (niżej też)

File: research/conditional/train/configs/baselines/gpt_small.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  lr_decay_interval: 1
+
+  logging_interval_heavy: 5000
+  logging_interval_loss: 1000
+  save_weights_path: "weights"
+  save_weights_interval: 1000
+
+  use_neptune: true
+  project_name: "pmtest/llm-random"
+  name: "gpt_medium_baseline"
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/gpt_small.yaml:33]: gpt_small_*

File: research/conditional/train/configs/baselines/gpt_small.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+# train params
+params:
+  # model
+  model_type: "gpt"
+  dataset_type: "c4"
+  dmodel: 512
+  dff: 2048
+  n_att_heads: 8
+  n_blocks: 4
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/gpt_small.yaml:14]: powinieneś chyba zmienić liczbę bloków na 8, a liczbę głów na 4 ?

**Comment by sebastianjaszczur** [comment]
Dałem trochę komentarzy co do configów. Jaki batch size się mieści? i chyba chcemy ustalić dużo większą liczbę kroków?

File: research/conditional/train/configs/baselines/bert_medium.yaml (hunk 1 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/bert_medium.yaml:3]: Nie, to pomyłka. Ustawiłem teraz na 2 dni.

File: research/conditional/train/configs/baselines/gpt_small.yaml (hunk 2 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/gpt_small.yaml:38]: Done

File: research/conditional/train/configs/baselines/bert_mini.yaml (hunk 3 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/bert_mini.yaml:38]: Done

File: research/conditional/train/configs/baselines/bert_medium.yaml (hunk 4 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/bert_medium.yaml:37]: Done

File: research/conditional/train/configs/baselines/gpt_small.yaml (hunk 5 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/gpt_small.yaml:33]: Done

File: research/conditional/train/configs/baselines/gpt_small.yaml (hunk 6 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/gpt_small.yaml:14]: Jest dobrze, w bercie small jest 8 głów i 4 bloki.

**Comment by TomaszOdrzygozdz** [comment]
Maksymalny batch size na bert medium dense   z loss chungusem to 505.

File: research/conditional/train/configs/baselines/gpt_medium.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  # model
+  model_type: "gpt"
+  dataset_type: "c4"
+  dmodel: 512
+  dff: 2048
+  n_att_heads: 8
+  n_blocks: 8
+  cutoff: 128
+  batch_size: 256
+  loss_checkpoint_chungs: 64
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt_medium.yaml:18]: How did this number come up? is it related to the vocab size somehow?

File: research/conditional/train/configs/baselines/gpt_small.yaml [hunk 8]
```diff
@@ -0,0 +1,39 @@
+runner: "research.conditional.train.cc_train"
+time: "2-00:00:00"
+n_gpus: 1
+interactive_debug: true
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt_small.yaml:4]: this should be false

File: research/conditional/train/configs/baselines/gpt_medium.yaml (hunk 7 above)
Comment (TomaszOdrzygozdz) [reply research/conditional/train/configs/baselines/gpt_medium.yaml:18]: This number should be around vocab_size / d_model.  64 is the closest that also divides the batch size.

File: research/conditional/train/configs/baselines/gpt_small.yaml (hunk 6 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/gpt_small.yaml:14]: potwierdzam wersję Tomka

File: research/conditional/train/configs/baselines/gpt_medium.yaml (hunk 7 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/gpt_medium.yaml:18]: I understand, thnanks

**Comment by Simontwice** [comment]
LGTM

**Comment by sebastianjaszczur** [comment]
A jaka jest reasonable liczba kroków?

File: research/conditional/train/configs/baselines/gpt_medium.yaml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+# train params
+params:
+  # model
+  model_type: "gpt"
+  dataset_type: "c4"
+  dmodel: 512
+  dff: 2048
+  n_att_heads: 8
+  n_blocks: 8
+  cutoff: 128
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/gpt_medium.yaml:16]: cutoff should be 256.
I've just became a fan of @Simontwice 's inheritance actually. But let's merge it without it.

File: research/conditional/train/configs/baselines/gpt_medium.yaml [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# grid args
+runner: "research.conditional.train.cc_train"
+time: "0-02:00:00"
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/gpt_medium.yaml:3]: Wciąż nie widzę 2d (dotyczy wszystkich plików)?

File: research/conditional/train/configs/baselines/gpt_medium.yaml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  n_blocks: 8
+  cutoff: 128
+  batch_size: 256
+  loss_checkpoint_chungs: 64
+
+  n_steps: 500000
+  grad_clip: 0.5
+  learning_rate: 0.0001
+  lr_decay: 0.9999777
+  lr_warmup_steps: 7500
```
Comment (sebastianjaszczur) [inline research/conditional/train/configs/baselines/gpt_medium.yaml:24]: chyba lepiej byłoby zmienić na 5k lub 10k (wszędzie) wygląda mniej arbitralnie. Ustalimy na spotkaniu.

File: research/conditional/train/configs/baselines/gpt_medium.yaml (hunk 9 above)
Comment (sebastianjaszczur) [reply research/conditional/train/configs/baselines/gpt_medium.yaml:16]: But target inheritance would define the whole dataset (and maybe even training steps, grad_clip, etc.) in the outermost file. Then dense baselines would inherit from those, adding their own model sizes etc. Then our ablations/custom models would inherit from dense baselines.

Comment (Simontwice) [reply research/conditional/train/configs/baselines/gpt_medium.yaml:16]: Yeah that's what Ive been thinking as well.

**Comment by Simontwice** [comment]
ten PR wisi za długo; baseliny nie mogą dłużej czekać. Ustawiłem lokalnie warmup na 10K, zapisuję modele co 50K, trenuję 250K+20K=270K kroków, zmieniłem nazwy ścieżek bo by się nadpisywały.

**Comment by maciek-pioro** [comment]
Another PR will be made with inheritance
